src/util/kaldi-table-inl.h

```diff
@@ -1631,25 +1748,48 @@ class RandomAccessTableReaderScriptImpl:
         return true;  // we have the key.
       } else {  // preload specified, so we have to pre-load the object before
               // returning true.
-        if (!input_.Open(script_[key_pos].second)) {
-          KALDI_WARN << "Error opening stream "
-                     << PrintableRxfilename(script_[key_pos].second);
-          return false;
+        std::string data_rxfilename; // We will split filename 
+        // script_[key_pos].second into data_rxfilename (e.g. 1.ark:100) 
+        // and range_(if any, e.g. [1:2,2:10]).
+        
+        if (script_[key_pos].second[script_[key_pos].second.size()-1] == ']') {
+          if(!ExtractRangeSpecifier(script_[key_pos].second, &data_rxfilename,
+                                    &range_)) {
+            KALDI_WARN << "TableReader: fail to parse file name "
+                       << PrintableRxfilename(script_[key_pos].second);
+            return false;
+          }
         } else {
-          // Make sure holder empty.
-          if (state_ == kHaveObject || state_ == kGaveObject)
-            holder_.Clear();
-          if (holder_.Read(input_.Stream())) {
-            state_ = kHaveObject;
-            current_key_ = key;
-            return true;
-          } else {
-            KALDI_WARN << "Error reading object from "
-                "stream " << PrintableRxfilename(script_[key_pos].second);
-            state_ = kNotHaveObject;
+          data_rxfilename = script_[key_pos].second;
+          range_ = "";
+        }
+        // we only need to update holder_ if data_rxfilename is different
+        // from last time.
+        if (data_rxfilename != data_rxfilename_last_) {
```

Comment:
oh, I see.  The data_rxfilename_ is not stored.  The variable is needed, but I think it should be called data_rxfilename_, not data_rxfilename_last_, because it relates to the currently held object, not the one before.